src/lib.rs

```diff
@@ -244,6 +262,148 @@ pub enum Credentials {
     /// Oauth client id and secret
     /// https://developer.github.com/v3/#oauth2-keysecret
     Client(String, String),
+    /// JWT token exchange, to be performed transparently in the
+    /// background. app-id, DER key-file.
+    /// https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/
+    JWT(JWTCredentials),
+    /// JWT-based App Installation Token
+    /// https://developer.github.com/apps/building-github-apps/authenticating-with-github-apps/
+    InstallationToken(InstallationTokenGenerator),
+}
+
+/// JSON Web Token authentication mechanism
+///
+/// The GitHub client methods are all &self, but the dynamically
+/// generated JWT token changes regularly. The token is also a bit
+/// expensive to regenerate, so we do want to have a mutable cache.
+///
+/// We use a token inside a Mutex so we can have interior mutability
+/// even though JWTCredentials is not mutable.
+#[derive(Debug, Clone)]
+pub struct JWTCredentials {
+    pub app_id: u64,
+    /// DER RSA key. Generate with
+    /// `openssl rsa -in private_rsa_key.pem -outform DER -out private_rsa_key.der`
```

Comment:
These private keys can't be generated, they can only be generated by GitHub and downloaded.

Reply:
That command doesn't generate the key ex nihilo, it just extracts ("generates") the binary key from the PEM file. Bad choice of word perhaps.